File: bench/cpu/store_forwarding.cpp
#include <benchmark/benchmark.h>

#include "common/array.h"

#include <atomic>

struct Foo1 {
  int32_t part1{0};
  int32_t part2{0};

  void Store(int32_t x) {
    // One of these must be correct, tested with both
    part1 = 0;
    part2 = x;

    // part2 = 0;
    // part1 = x;
  }

  int64_t Load() {
    return *reinterpret_cast<int64_t*>(this);
  }
};

struct Foo2 {
  int32_t part1{0};
  int32_t part2{0};

  void Store(int32_t x) {
    *reinterpret_cast<int64_t*>(this) = x;
  }

  int64_t Load() {
    return *reinterpret_cast<int64_t*>(this);
  }
};

template <typename Foo>
static void BM_StoreForwarding(benchmark::State& state) {
  auto arr = MakeArr<Foo>(1'000'000);
  volatile uint16_t x;

  int64_t sum{0};
  for (auto _ : state) {
    for (auto &elem : arr) {
      elem.Store(x);
      std::atomic_thread_fence(std::memory_order_seq_cst);
      sum += elem.Load();
      benchmark::DoNotOptimize(sum);
    }
  }
}

BENCHMARK(BM_StoreForwarding<Foo1>);
BENCHMARK(BM_StoreForwarding<Foo2>);

BENCHMARK_MAIN();

File: bench/cpu/store_forwarding.zig
const std = @import("std");

const N = 1_000_000;

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

const Elem = extern struct {
    part1: i32 = 0,
    part2: i32 = 0,
};

inline fn compilerFence() void {
    asm volatile ("" ::: "memory");
}

pub fn main() !void {
    const alloc = std.heap.page_allocator;
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    const arr = try alloc.alloc(Elem, N);
    defer alloc.free(arr);
    @memset(arr, .{});

    const x: i32 = 1;

    // Partial store then wide load — may fail store forwarding
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i64 = 0;
            for (arr) |*elem| {
                elem.part1 = 0;
                elem.part2 = x;
                compilerFence();
                sum +%= @as(*align(1) const i64, @ptrCast(elem)).*;
                std.mem.doNotOptimizeAway(sum);
            }
            total += timer.read();
        }
        try printResult("BM_StoreForwarding<Partial>", total, iters);
    }

    // Single wide store then wide load — store forwarding succeeds
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i64 = 0;
            for (arr) |*elem| {
                @as(*align(1) i64, @ptrCast(elem)).* = x;
                compilerFence();
                sum +%= @as(*align(1) const i64, @ptrCast(elem)).*;
                std.mem.doNotOptimizeAway(sum);
            }
            total += timer.read();
        }
        try printResult("BM_StoreForwarding<Full>", total, iters);
    }
}
